packages/babel-plugin-proposal-decorators/src/transformer-2021-12.ts

```diff
@@ -0,0 +1,987 @@
+import type { NodePath } from "@babel/traverse";
+import { types as t } from "@babel/core";
+import syntaxDecorators from "@babel/plugin-syntax-decorators";
+import ReplaceSupers from "@babel/helper-replace-supers";
+import * as charCodes from "charcodes";
+
+type ClassDecoratableElement =
+  | t.ClassMethod
+  | t.ClassPrivateMethod
+  | t.ClassProperty
+  | t.ClassPrivateProperty
+  | t.ClassAccessorProperty;
+
+type ClassElement =
+  | ClassDecoratableElement
+  | t.TSDeclareMethod
+  | t.TSIndexSignature
+  | t.StaticBlock;
+
+type classUidGenerator = <B extends boolean>(
+  isPrivate: B,
+) => B extends true ? t.PrivateName : t.Identifier;
+
+function incrementId(id: number[], idx = id.length - 1): void {
+  // If index is -1, id needs an additional character, unshift A
+  if (idx === -1) {
+    id.unshift(charCodes.uppercaseA);
+    return;
+  }
+
+  const current = id[idx];
+
+  if (current === charCodes.uppercaseZ) {
+    // if current is Z, skip to a
+    id[idx] = charCodes.lowercaseA;
+  } else if (current === charCodes.lowercaseZ) {
+    // if current is z, reset to A and carry the 1
+    id[idx] = charCodes.uppercaseA;
+    incrementId(id, idx - 1);
+  } else {
+    // else, increment by one
+    id[idx] = current + 1;
+  }
+}
+
+/**
+ * Generates a new element name that is unique to the given class. This can be
```

Comment:
"unique to the given class" might not be enough. It must be unique in the given class and in all the class enclosing the current one. For example:

```js
class A {
  #A = 1;
  static B = class B extends A {
    accessor a = 2;
    
    getA() {
      return this.#A;
    }
}

new A.B().getA();
```

we don't want to transform it to
```js
class A {
  #A = 1;
  static B = class B extends A {
    #A = 2;
    get a() { return this.#A }
    set a(v) { this.#A = v }
    
    getA() {
      return this.#A;
    }
}

new A.B().getA();
```
otherwise it returns `2` rather than `1`.

---

Now that this PR is mostly finished, could you add new commits without squashing :pray: It makes it easier to see what changed since the last review.

Reply:
Ok, updated the logic to remove the public UID part because we no longer use that, and now it takes in all referenced PrivateNames within the class body, so it should be sufficiently unique. If a duplicate key does exist within a parent, I don't think it should matter in the child unless the child references it, correct me if I'm wrong there.

> Now that this PR is mostly finished, could you add new commits without squashing 🙏 It makes it easier to see what changed since the last review.

No prob!